File: 94.BinaryTreeInorderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        inorderTraversal = []
        stack, ptr = [], root
        while stack or ptr:
            while ptr:
                stack.append(ptr)
                ptr = ptr.left
            ptr = stack.pop()
            inorderTraversal.append(ptr.val)
            ptr = ptr.right
        return inorderTraversal

File: 94.BinaryTreeInorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> inorderNum;
        stack<TreeNode*> visitedNodes;
        while (!visitedNodes.empty() || root) {
            while (root) {
                visitedNodes.push(root);
                root = root->left;
            }
            if (!visitedNodes.empty()) {
                auto node = visitedNodes.top();
                visitedNodes.pop();
                
                inorderNum.push_back(node->val);
                root = node->right;
            }
        }
        return inorderNum;
    }
};